### PR TITLE
(4.x) Merge 3.4

### PR DESCRIPTION
**Pull request by alalek**
#20511 from wjj19950828:add_humanseg_support_0806
#20573 from rogday:onnx_scale_fix
#20618 from VadimLevin:dev/vlevin/fix-vector-conversion
#20622 from diablodale:fix20544-nodiscard
#20633 from alalek:ocl_dumpValue_handle_null
#20635 from alalek:issue_20559
#20646 from VadimLevin:dev/vlevin/fix-vector-conversion

Previous "Merge 3.4": #20626

<cut/>

<details>

```
buildworker:Win64 OpenCL=windows-2
#buildworker:Custom=linux-1,linux-2,linux-4,linux-6
buildworker:Docs=linux-4,linux-6
build_image:Docs=docs-js:18.04
build_image:Custom=javascript
buildworker:Custom=linux-4,linux-6
Xbuild_image:Custom=javascript-simd
#build_image:Custom=powerpc64le
#build_image:Custom=ubuntu-openvino-2019r3.0:16.04
#build_image:Custom=ubuntu-openvino-2020.3.0:16.04
#build_image:Custom=ubuntu-openvino-2020.4.0:16.04
#build_image:Custom=ubuntu-openvino-2021.1.0:20.04
#build_image:Custom=ubuntu-openvino-2021.2.0:20.04
#build_image:Custom=ubuntu-openvino-2021.3.0:20.04
#build_image:Custom=ubuntu-openvino-2021.4.0:20.04
#buildworker:Custom=linux-1
#build_image:Custom=ubuntu-vulkan:16.04
#buildworker:Custom=linux-4
#build_image:Custom=fedora:28
#build_image:Custom=ubuntu-cuda:16.04
#build_image:Custom=ubuntu-clang:18.04
#build_image:Custom=ubuntu:20.04
#buildworker:Custom=linux-1
#build_image:Custom=javascript-simd
#build_image:Custom=mips64el
#build_image:Custom Mac=openvino-2019r3.0
#build_image:Custom Mac=openvino-2020.3.0
Xbuild_image:Custom Mac=openvino-2020.4.0
Xbuild_image:Custom Mac=openvino-2021.1.0
Xbuild_image:Custom Mac=openvino-2021.2.0
Xbuild_image:Custom Mac=openvino-2021.3.0
build_image:Custom Mac=openvino-2021.4.0
#build_image:Custom Win=openvino-2019r3.0
#build_image:Custom Win=openvino-2020.3.0
Xbuild_image:Custom Win=openvino-2020.4.0
Xbuild_image:Custom Win=openvino-2021.1.0
Xbuild_image:Custom Win=openvino-2021.2.0
Xbuild_image:Custom Win=openvino-2021.3.0
build_image:Custom Win=openvino-2021.4.0
test_bigdata:Custom Win=1
test_filter:Custom Win=*
test_modules=dnn,python2,python3,java
test_opencl:Custom Win=OFF
build_contrib:Custom Win=OFF
#build_image:Custom Win=msvs2017
#build_image:Custom Win=msvs2019
test_modules:Custom Mac=dnn,java,python3
```

</details>

### CODE REVIEW

File: modules/dnn/misc/python/test/test_dnn.py [hunk 1]
```diff
@@ -218,7 +218,7 @@ def test_textdetection_model(self):
         model.setInputParams(scale, size, mean)
         out, _ = model.detect(frame)
 
-        self.assertTrue(type(out) == list)
+        self.assertTrue(type(out) == tuple, msg='actual type {}'.format(str(type(out))))
```
Comment (alalek) [inline modules/dnn/misc/python/test/test_dnn.py:221]: This extra change is necessary after #20618 (list -> tuple)

/cc @VadimLevin

File: modules/gapi/misc/python/test/test_gapi_sample_pipelines.py [hunk 2]
```diff
@@ -517,8 +517,8 @@ def run(arr0, arr1):
 
             comp = cv.GComputation(cv.GIn(g_arr0, g_arr1), cv.GOut(g_out))
 
-            arr0 = [(2, 2), 2.0]
-            arr1 = [3,    'str']
+            arr0 = ((2, 2), 2.0)
+            arr1 = (3,    'str')
```
Comment (alalek) [inline modules/gapi/misc/python/test/test_gapi_sample_pipelines.py:521]: This extra change is necessary after #20618 (list -> tuple)

(or change `self.assertEqual(arr0 + arr1, out)` below)

/cc @VadimLevin

File: modules/core/include/opencv2/core/mat.hpp [hunk 3]
```diff
@@ -2451,11 +2451,11 @@ class CV_EXPORTS UMat
     //!  <0 - a diagonal from the lower half)
     UMat diag(int d=0) const;
     //! constructs a square diagonal matrix which main diagonal is vector "d"
-    static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-    static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
+    CV_NODISCARD_STD static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
+    CV_NODISCARD_STD static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
```
Comment (alalek) [inline modules/core/include/opencv2/core/mat.hpp:2455]: Resolved `CV_NODISCARD_STD` conflicts from patch #20622

/cc @diablodale 

BTW, why nodiscard flag is missing for methods above, like

`UMat diag(int d=0) const;`

**Review by alalek — COMMENTED** [review]
/cc @VadimLevin @diablodale Please take a look on this merge to OpenCV 4.x branch

File: modules/gapi/misc/python/test/test_gapi_sample_pipelines.py (hunk 2 above)
Comment (VadimLevin) [reply modules/gapi/misc/python/test/test_gapi_sample_pipelines.py:521]: Maybe: `self.assertSequenceEqual(arr0 + arr1, out)` would be better. Example:
```python
        from itertools import chain
        a = [1, 2, [1, 2]]
        b = ['str', (2, 2)]
        res = tuple(chain(a, b))
        b[0] = 'a' # This modification is required to fail assertion below
        self.assertSequenceEqual(a + b, res)
```
reports

> AssertionError: Sequences differ: [1, 2, [1, 2], 'a', (2, 2)] != (1, 2, [1, 2], 'str', (2, 2))
> 
> First differing element 3:
> 'a'
> 'str'
> 
> - [1, 2, [1, 2], 'a', (2, 2)]
> + (1, 2, [1, 2], 'str', (2, 2))

**Comment by VadimLevin** [comment]
@alalek  Should #20646 be included to this merge too from the beginning?

File: modules/core/include/opencv2/core/mat.hpp (hunk 3 above)
Comment (diablodale) [reply modules/core/include/opencv2/core/mat.hpp:2455]: Not having CV_NODISCARD_STD on `UMat diag(int d=0)` was a style choice. `diag` is a member function, not a static factory. Static factories have the "feature" that they can be called using `class::factory` or `var.factory` which contributes to discarding errors.

There are hundreds/thousands of APIs in opencv that create an object and return it. Technically... nodiscard could be put on all of them. But that is overkill and not really the intention of [[nodiscard]]. For this first PR effort, I focused on the static factories that create new independent objects. There may be future opportunities on some constructors (like those that do locking), but that wasn't the focus on this first effort. For example, I could put nodicard on the UMat constructors, but it is harder to make a discarding error with UMat constructor...compared to a discarding error with UMat::zeros

The one exception was member functions `clone()`. Those were already marked and therefore I want to maintain the same functionality on them.

Comment (diablodale) [reply modules/core/include/opencv2/core/mat.hpp:2455]: Also... I don't visually 👀 see any differences between my original nodiscard commit and this PR's changes.
You wrote you resolved conflicts. Is there a way in github to see the specific changes you applied to resolve conflicts?

File: modules/core/include/opencv2/core/mat.hpp [hunk 4]
```diff
@@ -2486,22 +2486,22 @@ class CV_EXPORTS UMat
...
-    static UMat ones(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-    static UMat ones(int rows, int cols, int type) { return ones(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-    static UMat ones(Size size, int type) { return ones(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-    static UMat ones(int ndims, const int* sz, int type) { return ones(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-    static UMat eye(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-    static UMat eye(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-    static UMat eye(int rows, int cols, int type) { return eye(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-    static UMat eye(Size size, int type) { return eye(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
+    CV_NODISCARD_STD static UMat zeros(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
```
Comment (diablodale) [inline modules/core/include/opencv2/core/mat.hpp:2489]: This block of merge changes looks ok. After merge, I will review/make PR on the 4.x codebase looking for static factories that are unique to 4.x.

File: modules/core/include/opencv2/core/mat.hpp (hunk 3 above)
Comment (diablodale) [reply modules/core/include/opencv2/core/mat.hpp:2455]: this block merge changes looks ok also

Comment (alalek) [reply modules/core/include/opencv2/core/mat.hpp:2455]: `git show <merge commit sha>` can show conflicting changes:

<details>

```.diff
commit 5aa7435d2537e677611390b530e31b3b204773a3 (HEAD -> merge-3.4, github/test_contrib, github/merge-3.4)
Merge: c11195d5e3 7c23ec90a9
Author: Alexander Alekhin <alexander.a.alekhin@gmail.com>
Date:   Thu Sep 2 15:24:04 2021 +0000

    Merge remote-tracking branch 'upstream/3.4' into merge-3.4

diff --cc modules/core/include/opencv2/core/mat.hpp
index 6768be7683,fdcb8fc817..1ef0eb5a02
--- a/modules/core/include/opencv2/core/mat.hpp
+++ b/modules/core/include/opencv2/core/mat.hpp
@@@ -2451,11 -2469,10 +2451,11 @@@ public
      //!  <0 - a diagonal from the lower half)
      UMat diag(int d=0) const;
      //! constructs a square diagonal matrix which main diagonal is vector "d"
-     static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
 -    CV_NODISCARD_STD static UMat diag(const UMat& d);
++    CV_NODISCARD_STD static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
  
      //! returns deep copy of the matrix, i.e. the data is copied
-     UMat clone() const CV_NODISCARD;
+     CV_NODISCARD_STD UMat clone() const;
      //! copies the matrix content to "m".
      // It calls m.create(this->size(), this->type()).
      void copyTo( OutputArray m ) const;
@@@ -2486,22 -2503,14 +2486,22 @@@
      double dot(InputArray m) const;
  
      //! Matlab-style matrix initialization
-     static UMat zeros(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat zeros(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat zeros(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat zeros(int rows, int cols, int type) { return zeros(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat zeros(Size size, int type) { return zeros(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat zeros(int ndims, const int* sz, int type) { return zeros(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat ones(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat ones(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat ones(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat ones(int rows, int cols, int type) { return ones(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat ones(Size size, int type) { return ones(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat ones(int ndims, const int* sz, int type) { return ones(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat eye(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat eye(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat eye(int rows, int cols, int type) { return eye(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat eye(Size size, int type) { return eye(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
 -    CV_NODISCARD_STD static UMat zeros(int rows, int cols, int type);
 -    CV_NODISCARD_STD static UMat zeros(Size size, int type);
 -    CV_NODISCARD_STD static UMat zeros(int ndims, const int* sz, int type);
 -    CV_NODISCARD_STD static UMat ones(int rows, int cols, int type);
 -    CV_NODISCARD_STD static UMat ones(Size size, int type);
 -    CV_NODISCARD_STD static UMat ones(int ndims, const int* sz, int type);
 -    CV_NODISCARD_STD static UMat eye(int rows, int cols, int type);
 -    CV_NODISCARD_STD static UMat eye(Size size, int type);
++    CV_NODISCARD_STD static UMat zeros(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat zeros(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat zeros(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat zeros(int rows, int cols, int type) { return zeros(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat zeros(Size size, int type) { return zeros(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat zeros(int ndims, const int* sz, int type) { return zeros(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat ones(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat ones(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat ones(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat ones(int rows, int cols, int type) { return ones(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat ones(Size size, int type) { return ones(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat ones(int ndims, const int* sz, int type) { return ones(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat eye(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat eye(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat eye(int rows, int cols, int type) { return eye(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat eye(Size size, int type) { return eye(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
  
      //! allocates new matrix data unless the matrix already has specified size and type.
      // previous data is unreferenced if needed.
diff --cc modules/core/include/opencv2/core/matx.hpp
index 45d8a97f34,f25c8bce57..3c92e3a21d
--- a/modules/core/include/opencv2/core/matx.hpp
+++ b/modules/core/include/opencv2/core/matx.hpp
@@@ -140,13 -142,15 +140,13 @@@ public
           _Tp v12, _Tp v13, _Tp v14, _Tp v15); //!< 1x16, 4x4 or 16x1 matrix
      explicit Matx(const _Tp* vals); //!< initialize from a plain array
  
 -#ifdef CV_CXX11
      Matx(std::initializer_list<_Tp>); //!< initialize from an initializer list
 -#endif
  
-     static Matx all(_Tp alpha);
-     static Matx zeros();
-     static Matx ones();
-     static Matx eye();
-     static Matx diag(const diag_type& d);
+     CV_NODISCARD_STD static Matx all(_Tp alpha);
+     CV_NODISCARD_STD static Matx zeros();
+     CV_NODISCARD_STD static Matx ones();
+     CV_NODISCARD_STD static Matx eye();
+     CV_NODISCARD_STD static Matx diag(const diag_type& d);
      /** @brief Generates uniformly distributed random numbers
      @param a Range boundary.
      @param b The other range boundary (boundaries don't have to be ordered, the lower boundary is inclusive,
diff --cc modules/core/src/bindings_utils.cpp
...
```

</details>

**Comment by alalek** [comment]
:+1:

**Pull request merged**